Platform/AMD/AmdPlatformPkg/DynamicTables/Library/Acpi/AcpiSsdtPciLib/AcpiSsdtPciLib.c

```diff
@@ -0,0 +1,1540 @@
+             NULL,
+             (UINT32)(RootBridge->GlobalInterruptStart + 1),
+             PrtNode
+             );
+  ASSERT_EFI_ERROR (Status);
+
+  for (Index = 1; Index <= RootBridge->RootPortCount; Index++) {
+    if ((RootBridge->RootPort[Index]->PortPresent == 0) &&
+        (RootBridge->RootPort[Index]->Enabled == 0))
+    {
+      continue;
+    }
+
+    // Only insert for Functions 1 - 4 (minus 1)
+    if (((RootBridge->RootPort[Index]->Function - 1) & ~0x3) == 0) {
+      Status = AmlAddPrtEntry (
+                 (UINT32)((RootBridge->RootPort[Index]->Device << 16) | 0x0000FFFF),
+                 (UINT8)(RootBridge->RootPort[Index]->Function - 1),
+                 NULL,
+                 (UINT32)(RootBridge->GlobalInterruptStart + RootBridge->RootPort[Index]->EndpointInterruptArray[RootBridge->RootPort[Index]->Function - 1]),
+                 PrtNode
+                 );
+      ASSERT_EFI_ERROR (Status);
+    }
+  }
+
+  // Attach the _PRT entry.
+  Status = AmlAttachNode (PciNode, PrtNode);
+  if (EFI_ERROR (Status)) {
+    AmlDeleteTree (PrtNode);
+    ASSERT_EFI_ERROR (Status);
+  }
+
+  PrtNode = NULL;
+
+  if (EFI_ERROR (Status)) {
+    return EFI_DEVICE_ERROR;
+  }
+
+  return Status;
+}
+
+/**
+  Insert Root Bridge resources into AML table
+
+  @param[in]      RootBridge  - Single Root Bridge instance
+  @param[in, out] CrsNode     - AML tree node
+
+  @retval         EFI_SUCCESS, various EFI FAILUREs.
+**/
+EFI_STATUS
+EFIAPI
+InternalInsertRootBridgeResources (
+  IN      AMD_PCI_ROOT_BRIDGE_OBJECT_INSTANCE  *RootBridge,
+  IN OUT  AML_OBJECT_NODE_HANDLE               CrsNode
+  )
+{
+  EFI_ACPI_QWORD_ADDRESS_SPACE_DESCRIPTOR  *LocalBuffer;
+  EFI_HANDLE                               *HandleBuffer;
+  EFI_PCI_ROOT_BRIDGE_IO_PROTOCOL          *Io;
+  EFI_STATUS                               Status;
+  UINTN                                    BaseBusNumber;
+  UINTN                                    Index;
+  UINTN                                    NumHandles;
+  VOID                                     *Configuration;  // Never free this buffer
+
+  BaseBusNumber = ~(UINTN)0;
+  // Get EFI Pci Root Bridge I/O Protocols
+  Status = gBS->LocateHandleBuffer (
+                  ByProtocol,
+                  &gEfiPciRootBridgeIoProtocolGuid,
+                  NULL,
+                  &NumHandles,
+                  &HandleBuffer
+                  );
+  if (EFI_ERROR (Status)) {
+    return Status;
+  }
+
+  // Locate the Root Bridge IO protocol for this root bridge.
+  LocalBuffer   = NULL;
+  Configuration = NULL;
+  for (Index = 0; Index < NumHandles; Index++) {
+    Status = gBS->OpenProtocol (
+                    HandleBuffer[Index],
+                    &gEfiPciRootBridgeIoProtocolGuid,
+                    (VOID **)&Io,
+                    mDriverHandle,
+                    NULL,
+                    EFI_OPEN_PROTOCOL_GET_PROTOCOL
+                    );
+    if (EFI_ERROR (Status)) {
+      return Status;

```

Comment:
You may want to free HandleBuffer in all of the error conditions in this function.

Reply:
done.